### PR TITLE
Updated TypeScript to allow defined user data

### PR DESCRIPTION
**Pull request by biw**
The PR changes the `WebSocket` type definition to allow for an *optional* type argument that allows for strictly defined `WebSocket` user data.

Previously, when you extended the `WebSocket` object in TypeScript, you did not get the benefit of the typing:

```ts
interface ExtendWebSocket extends WebSocket {
  username: 'biw'
}

uWebSockets.App().ws('/*', {
  open: (ws: ExtendWebSocket) => {
    ws.username = 'biw'
   
    // no error for key not in ExtendWebSocket
    ws.token = '123'
  },
```

With this change, you have the *option* for more strict type checking:

```ts
interface ExtendValues {
  username: 'biw'
}

uWebSockets.App().ws('/*', {
  open: (ws: ExtendWebSocket<ExtendValues>) => {
    ws.username = 'biw'
   
    // will throw typescript error for undefined key
    ws.token = '123'
  },
```

The default behavior is still the same and non-breaking:
```ts
uWebSockets.App().ws('/*', {
  open: (ws) => {
    // no typescript errors for either
    ws.username = 'biw'
    ws.token = '123'
  },
```

Happy to make any changes you'd like, just let me know!

### CODE REVIEW

**Comment by alexhultman** [comment]
Interesting. Is it possible to have it be entirely like the C++ interface:

```
App().ws<T>({
open: (ws: WebSocket<T>) => {
T userData = ws.getUserData();
}
})
```

Implementation wise getUserData() would just return "this"

**Comment by alexhultman** [comment]
Alright, so it just needs getUserData function now, returning UserData. Is it possible to return a reference to UserData?

ws.getUserData().variable = 13;

**Comment by biw** [comment]
Oh, that's even better. At first, I misread your reply and I updated the types to match c++ while still being keys on the `WebSocket` object.

Not 100% sure of the C++ side, but if you pass an object around in JavaScript it will work since javascript passes all objects by reference.

**Comment by alexhultman** [comment]
Right you cannot have number as T, it had to be a complete object and then it's by reference either way

**Comment by biw** [comment]
Happy to contribute typescript types for the change to `ws.getUserData()` in Javascript or even writing the feature with a little guidance. I want to help, not waste your time, so good either way.

Do you want to move to `ws.getUserData()` in JS before we merge the current PR?

File: docs/index.d.ts [hunk 1]
```diff
@@ -13,7 +13,10 @@ export type RecognizedString = string | ArrayBuffer | Uint8Array | Int8Array | U
 /** A WebSocket connection that is valid from open to close event.
  * Read more about this in the user manual.
  */
-export interface WebSocket {
+export type WebSocket<UserData = {
```
Comment (yovanoc) [inline docs/index.d.ts:16]: I think it would be better here to do UseerData extends and not = if it only can be an object

Comment (biw) [reply docs/index.d.ts:16]: Sorry, can you please give an example of what you mean?

Comment (alexhultman) [reply docs/index.d.ts:16]: What about just making it ```WebSocket<UserData>```

Comment (yovanoc) [reply docs/index.d.ts:16]: When you do 

```ts
UserData = {
    /** Arbitrary user data may be attached to this object. */
    [key: string]: any;
}
```

You can pass any type you want, but you have this one by default but you can pass `WebSocket<number>` if you want and getUserData() will return a number for typescript. So I said that if its always an objet we have to write

```ts
UserData extends {
    /** Arbitrary user data may be attached to this object. */
    [key: string]: any;
}
```

to force the TS user to only pass an "object" like type

Comment (biw) [reply docs/index.d.ts:16]: Sorry, @yovanoc, do you mean: `export type WebSocket<UserData extends {`?

If we change it to `UserData extends {` or `WebSocket<UserData>`, the code then requires the type argument which would break the current exported API.

Comment (yovanoc) [reply docs/index.d.ts:16]: image

You can always put `= {}` at the end

Comment (yovanoc) [reply docs/index.d.ts:16]: image

Comment (alexhultman) [reply docs/index.d.ts:16]: Just throw away that key value thing, it doesn't belong there it was added by someone else. It should just take a type, preferably object

Comment (yovanoc) [reply docs/index.d.ts:16]: If getUserData was the only way to get data and it accept anything it would be perfect but that would cause a problem if we don't pass an object

image

Comment (biw) [reply docs/index.d.ts:16]: Ahh, good catch @yovanoc.

So, if we want to keep the same API (not requiring the typing out all UserData), then we should do:

```ts
UserData extends {
    /** Arbitrary user data may be attached to this object. */
    [key: string]: any;
} = { [key: string]: any } 
```


I can't think of a library in typescript that requires types, are you sure you want me to remove default types?

Comment (alexhultman) [reply docs/index.d.ts:16]: Problem with TypeScript is you lose the type information once it compiles to JS, so the template function ````ws<UserData>```` cannot properly call the constructors of UserData, because you don't know their type.

Regarding extends, isn't there some "extends Object" kind of thing?

Comment (alexhultman) [reply docs/index.d.ts:16]: The only way to make this interface work would be to wrap the entire API in a TypeScript wrapper that takes the type of UserData and makes sure to call the constructor/destructor of UserData before/after emitting open/close events.

And that sucks.

**Comment by Prozi** [comment]
seems like cache issue with CI 
```
warning: unable to rmdir 'uWebSockets': Directory not empty
```

**Comment by Prozi** [comment]
cc @alexhultman

**Comment by alexhultman** [comment]
I'm closing this for reasons discussed above. It cannot be easily mapped since a constructor in TypeScript is not a constructor in native land.

**Comment by pavulon** [comment]
Just wanted to clear up some possible confusion here. This PR never indented to make user data work the same way as in C++ where the generic type is actually instantiated (for obvious reasons mentioned above), but once the API got deceptively close it got derailed a bit. It was just trying to add a possibility to opt-in into some additional sanity-check errors and nicer autocomplete, which is basically all typescript can do for this project.

Anyway, the proposed solution wasn't 100% correct because there's no guarantee that your custom properties are initialized before you try using them, so everything should be marked as optional.

**Comment by alexhultman** [comment]
So what can be done here?

**Comment by alexhultman** [comment]
https://github.com/uNetworking/uWebSockets/issues/1160

I don't know if the first commit here can be merged after that change?

**Comment by biw** [comment]
> Just wanted to clear up some possible confusion here. This PR never intended to make user data work the same way as in C++ where the generic type is actually instantiated (for obvious reasons mentioned above), but once the API got deceptively close it got derailed a bit. It was just trying to add a possibility to opt-in into some additional sanity-check errors and nicer autocomplete, which is basically all typescript can do for this project.

Agreed, that was my original goal here. (sorry for the late response, didn't see this comment till now)

> Anyway, the proposed solution wasn't 100% correct because there's no guarantee that your custom properties are initialized before you try using them, so everything should be marked as optional.

Yes, that probably would be the correct solution, but couldn't that be left up to the application code to decide rather than the library?



> So what can be done here?

I think I may be missing something, but I think the first two commits can be merged without introducing API changes.

**Comment by alexhultman** [comment]
Yes if you roll back to first 2 commits it should be fine for v19

**Comment by biw** [comment]
Perfect, will do

**Comment by Dealerpriest** [comment]
Hello there! I was already initiating a question/discussion in the commit that reverted this merge. But I realized it might be more appropriate to write here 👍

I'm curious of whether this PR was perhaps too quickly dissmissed? From discussions it seems to me that the argument against this commit is that typescript loses typesafety when compiled to JS and this would in some way be problematic. I fail to see/understand that reasoning, as typescripts main role is to help with typechecking during development before compiled to JS, and the compiled JS would (to the best of my knowledge) in this case not differ in a way that mattered.
The purpose of the PR is pretty well explained by @pavulon's comment above, and then further clarified by @biw. And the consensus then seemed to me to be that the PR in fact made sense?

The JS api conveniently already allows custom user data to be provided, but unfortunately without any typescript type checking or info about the shape/type of the user data. This makes it more error prone when interacting with the user data in TS/JS land. For example writing `userName` instead of `username` wouldn't be catched by the type checker since it doesn't expect any particular shape of the data. This PR keeps the default (very wide) shape of the user data if no typescript generic is provided:
```
[key:string]: any
```
but allows the user to provide a shape trhough the type parameter. This helps the developer correctly use the content of the user data correctly inside the websocketBehaviour functions such as open and message.

If I'm misunderstanding, I would be grateful if someone helped me understand why this PR should not be merged👍

Here is an image that shows the nice in-editor type feedback this PR gives me:
image

**Comment by uNetworkingAB** [comment]
> I'm closing this for reasons discussed above. It cannot be easily mapped since a constructor in TypeScript is not a constructor in native land.

The problem is that constructors of T cannot be called, since the type information is lost once it goes from Typescript to JavaScript.

Uws is native only and sees only JavaScript so when you call app.ws<T> that should have different runtime behavior than app.ws<Q>, they are two different constructors. But uws only sees it as two calls to app.ws, there is no type information passed down.

So implementing this interface cannot be properly done

**Comment by uNetworkingAB** [comment]
You need some kind of typescript shim layer that would translate app.ws of T (...) to app.ws(T.constructor, ...) But that's very intrusive and goes way beyond the time budget for this.

I like the interface, that's the interface in C++ but it would break typescript constructors

**Comment by uNetworkingAB** [comment]
Does typescript have non constructible types? Interface cannot have a constructor?

If we can guarantee T cannot be something with a default constructor it can work

**Comment by uNetworkingAB** [comment]
TLDR - if you can make it the same interface as in C++ (guaranteeing no constructors or default inits for T) then I'm all for making it entirely static, only. No opt-in just full on typed. But I have no idea how typescript works so you have to present a case with all the parts working

**Comment by Dealerpriest** [comment]
In typescript an interface corresponds to a plain javascript object and describes the shape of the object and it's types. Thus a typescript interface can't have a constructor. 

With typescript the types are only ever used in the editor or (sometimes) during transpilation to javascript (but, to be clear, some transpilers like esbuild for example, dont care about the types at all and just strip them away from the code).

Given the above paragraph, I don't believe this PR can in any way affect how the user data is handled inside uws during runtime. Is the issue perhaps more related to the consuming side? The library user that is, and how the exported typescript types affect what the users will provide when calling the upgrade function? 
Im trying to digest and understand exactly what you have written. If I understand correctly you're after something that restricts which types could be passed to the generic so that it can't receive anything that would need "constructing". The closest to constructor would be a class, function or an object with functions I think? 
Perhaps then narrow down the allowed type passed to the generic by doing something like this:
```
ws<T extends Record<string, Object | string | number | etc>>()
```

Sorry if I'm in the wrong or mistaken! I'm truly trying to get a better understanding. I have a feeling it still might cause some confusion with the similarity between the c++ templating and the typescript generics. They look so similar, but (to the best of my knowledge) works completely different.

**Comment by uNetworkingAB** [comment]
Sounds good can you make a PR let's just go for interface T and also I don't want WebSocket to be a type, keep it interface. All the things should be interface
